src/sink.rs

```diff
@@ -0,0 +1,113 @@
+
+use crate::Ratelimit;
+use futures::{Async, AsyncSink, Future, Poll, Sink, StartSend, Stream};
+use ratelimit_meter::{algorithms::Algorithm, DirectRateLimiter, NonConformance};
+use std::io;
+
+pub trait SinkExt: Sink {
+    /// Limits the rate at which items can be put into the current sink.
+    ///
+    /// Note that this combinator does not buffer any items internally but
+    /// instead limits `start_send` calls to the underlying sink.
+    fn sink_ratelimit<A: Algorithm>(self, limiter: DirectRateLimiter<A>) -> Ratelimited<Self, A>
+    where
+        Self: Sized,
+        <A as Algorithm>::NegativeDecision: NonConformance;
+}
+
+impl<S: Sink> SinkExt for S {
+    fn sink_ratelimit<A: Algorithm>(self, limiter: DirectRateLimiter<A>) -> Ratelimited<Self, A>
+    where
+        Self: Sized,
+        <A as Algorithm>::NegativeDecision: NonConformance,
+    {
+        Ratelimited {
+            inner: self,
+            ratelimit: Ratelimit::new(limiter),
+            check_limit: true,
+        }
+    }
+}
+
+/// A sink combinator which will limit the rate of items passing through.
+///
+/// This is produced by the [SinkExt::sink_ratelimit] methods.
+pub struct Ratelimited<S: Sink, A: Algorithm>
+where
+    <A as Algorithm>::NegativeDecision: NonConformance,
+{
+    inner: S,
+    ratelimit: Ratelimit<A>,
+    check_limit: bool,
+}
+
+impl<S: Sink, A: Algorithm> Ratelimited<S, A>
+where
+    <A as Algorithm>::NegativeDecision: NonConformance,
+{
+    /// Acquires a references to the underlying sink that this combinator is pushing to.
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Acquires a mutable references to the underlying sink that this combinator is pushing to.
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consumes this combinator, returning the underlying sink.
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+
+impl<S: Sink + Stream, A: Algorithm> Stream for Ratelimited<S, A>
+where
+    <A as Algorithm>::NegativeDecision: NonConformance,
+{
+    type Item = S::Item;
+    type Error = S::Error;
+
+    fn poll(&mut self) -> Poll<Option<S::Item>, S::Error> {
+        self.inner.poll()
+    }
+}
+
+impl<S: Sink, A: Algorithm> Sink for Ratelimited<S, A>
+where
+    <A as Algorithm>::NegativeDecision: NonConformance,
+    <S as Sink>::SinkError: From<io::Error>,
+{
+    type SinkItem = S::SinkItem;
+    type SinkError = S::SinkError;
+
+    fn start_send(&mut self, item: S::SinkItem) -> StartSend<S::SinkItem, S::SinkError> {
+        if self.check_limit {
+            // Wait until we're good to go
+            if let Async::NotReady = self.ratelimit.poll()? {
+                return Ok(AsyncSink::NotReady(item));
+            }
+            self.check_limit = false;
+        }
+
+        // Try to push the item into underlying sink
+        if let AsyncSink::NotReady(item) = self.inner.start_send(item)? {
+            return Ok(AsyncSink::NotReady(item));
```

Comment:
I think there might be a weird interaction here: If the limiter says the stream can send, and the underlying stream isn't ready, we'll wait until ready. But if more than one thing does this for a longer time, there might be some "bunching up" of items as more things try to send into it than the rate limit allows.

I'm not super familiar with futures 0.1 streams, is there a way to check if the underlying stream is ready to receive without actually sending an item? (I think partial readiness might be a problem here... in fact, it would be interesting to consider the case of "leases" in the rate-limiter, indicating that we're reserving a certain amount of work?

Reply:
I think you're having streams and sinks mixed up a bit.
In the `futures` crate, `Stream`s produce items when you `poll` them whereas `Sink`s consume items when you `send` an item into a sink.
For now I'm going to assume you meant "sink" wherever you said "stream" since that's the code you commented on (though I think there may be a similar interaction when you have multiple streams; PS: yes there is, see last two paragraphs).

No, there's no way to check if a sink is full without sending it an item.

I don't think something like leases in the rate-limiter can work because there's no contract in the Sink trait which would require a call to `start_send` (where the combinator could acquire a lease from the limiter) to be followed by another call to `start_send` (e.g. If some real-time-relevant piece of data keeps getting rejected, at some point it might no longer make sense to try to send it). So there's no guarantee we will actually be able to do anything with the lease after we've acquired it and we might end up holding onto it indefinitely.

I'm not even sure if limiting the rate at which the inner sink accepts items is a valid goal (that's what the current implementation tries to do). For all we know, it may be buffering the items and then sending them in batches which could violate the limiter anyway.
Instead, one viable goal would be limiting the rate at which items can be sent into the combinator itself. To some extend that's what the current implementation does but, as you noticed, it falls short when there are multiple combinators using the same limiter.

Having re-read the documentation on the Sink trait, I think the correct wait to fix this would be introducing a single-element buffer to the combinator. `start_send` can then check the readiness of this buffer before querying the limiter. As such, it can (and should) always immediately consume the given item (by storing it in the buffer) when the limiter gives the OK.
The item in the buffer will then only be passed to the inner sink when `poll_complete` is called.
This need to call `poll_complete` already is a general contract of the Sink trait anyway:
> **You *must* use `poll_complete` in order to drive completion of a send**. In particular, `start_send` does not begin the flushing process

For streams the two analogous goals are:
1. Limit the rate at which the inner stream is polled / can produce items (current implementation)
2. Limit the rate at which the combinator will produce items
This would be similarly implemented by always polling the inner stream while a single-element buffer is empty and only when it is not empty is the limiter queried and an item produced when it returns the OK.

Option 1 suffers from the same "bunching up" as the current sink combinators do when you have multiple stream combinators all using the same limiter.
I think targeting option 2 is the correct thing to do here as well for all the same reasons.

Reply:
That makes sense - I clearly have a bunch to read up on re. streams/sinks. I'll do so when I come back from my vacation & get this merged then! (By Friday/Saturday, I think)